### PR TITLE
Adds support for custom asynchronous functions

### LINKED ISSUE #40: Async functions
Awesome job!

Will Power-Fx support asynchronous functions?

### PR DESCRIPTION
**Pull request by jas-valgotar**
-Fixes #40 
-Adds capability to register custom async function with Power FX.

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by MikeStall** [comment]
>     public class CustomFunctions : PowerFxTest

Another test:
Custom functions can also bind to services from the ServiceProvider. 

See this test:
[https://github.com/microsoft/Power-Fx/blob/9330cd645ca5a9bcc7368d5d407022b4fec0db03/src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs#L379](https://github.com/microsoft/Power-Fx/blob/9330cd645ca5a9bcc7368d5d407022b4fec0db03/src/tests/Microsoft.PowerFx.Interpreter.Tests/ConfigTests.cs#L379)

The 1st arg is actually bound to services from the RuntimeConfig.  
 public StringValue Execute(Runtime config, NumberValue x)

Can we make this async, ala:
public Task<StringValue> Execute(Runtime config, NumberValue x)




---
Refers to: src/tests/Microsoft.PowerFx.Interpreter.Tests/CustomFunctions.cs:16 in 01818b3. [](commit_id = 01818b3b86ebc5c6b415bf4ce0ed40acf1cdb898, deletion_comment = False)

File: src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs [hunk 1]
```diff
@@ -262,6 +304,14 @@ internal TexlFunction GetTexlFunction()
...
+            {
+                return new CustomAsyncTexlFunction(info.Name, info.RetType, info.ParamTypes)
+                {
+                    _impl = (args, cancellationToken) => InvokeAsync(args, cancellationToken)
+                };
+            }
+
             return new CustomTexlFunction(info.Name, info.RetType, info.ParamTypes)
             {
                 _impl = (runtimeConfig, args) => Invoke(runtimeConfig, args)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs:317]: >runtimeConfig [](http://example.com/codeflow?start=56&length=13)

We'll also need to pass this in the async case.

**Comment by MikeStall** [comment]
>         public virtual FormulaValue Invoke(IServiceProvider serviceProvider, FormulaValue[] args)

Can we just make this function async? 
It's ok if it calls a sync call that doesn't take advantage of being async. 

And then avoid adding new class CustomAsyncTexlFunction ?



---
Refers to: src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs:53 in 01818b3. [](commit_id = 01818b3b86ebc5c6b415bf4ce0ed40acf1cdb898, deletion_comment = False)

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/CustomFunctions.cs [hunk 2]
```diff
@@ -38,6 +42,82 @@ public static StringValue Execute(NumberValue x, BooleanValue b)
...
+                var val = x.Value.ToString() + "," + b.Value.ToString();
+                return FormulaValue.New(val);
+            }
+        }
+
+        private class TestCustomWaitAsyncFunction : ReflectionFunction
+        {
+            private readonly TaskCompletionSource<FormulaValue> _waiter = new TaskCompletionSource<FormulaValue>();
+
+            public async Task<NumberValue> Execute()
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/CustomFunctions.cs:104]: another test: async functions should be able to model bind to the CancellationToken.

File: src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs [hunk 3]
```diff
@@ -127,6 +127,40 @@ public async Task<FormulaValue> InvokeAsync(FormulaValue[] args, CancellationTok
...
+            : this(name, returnType._type, Array.ConvertAll(paramTypes, x => x._type))
+        {
+        }
+
+        public CustomAsyncTexlFunction(string name, DType returnType, params DType[] paramTypes)
+            : base(DPath.Root, name, name, SG("Custom func " + name), FunctionCategories.MathAndStat, returnType, 0, paramTypes.Length, paramTypes.Length, paramTypes)
+        {
+        }
+
+        public override bool IsSelfContained => true;
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs:146]: Are these necessarily always side effect-free, or should we ask implementers to provide this in the constructor?

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:x: 1 Binary Compatibility issues for **Microsoft.PowerFx.Interpreter** compared against [version 0.2.3-preview.20220928-323377](https://www.nuget.org/packages/Microsoft.PowerFx.Interpreter/0.2.3-preview.20220928-323377).
<details>

```
MembersMustExist : Member 'Microsoft.PowerFx.ReflectionFunction.Invoke(System.IServiceProvider, Microsoft.PowerFx.Types.FormulaValue[])' does not exist in the implementation but it does exist in the contract.
```

</details>
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs [hunk 4]
```diff
@@ -258,17 +281,17 @@ internal TexlFunction GetTexlFunction()
...
 
             return new CustomTexlFunction(info.Name, info.RetType, info.ParamTypes)
             {
-                _impl = (runtimeConfig, args) => Invoke(runtimeConfig, args)
+                _impl = (runtimeConfig, args, cancellationToken) => InvokeAsync(runtimeConfig, args, cancellationToken)
             };
         }
 
-        public FormulaValue Invoke(IServiceProvider serviceProvider, FormulaValue[] args)
+        public async Task<FormulaValue> InvokeAsync(IServiceProvider serviceProvider, FormulaValue[] args, CancellationToken cancellationToken)
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs:294]: >InvokeAsync [](http://example.com/codeflow?start=40&length=11)

This looks like the breaking change we're getting flagged on...

File: src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs [hunk 5]
```diff
@@ -207,21 +209,36 @@ private FunctionDescr Scan()
                 info.RetType = GetType(m.ReturnType);
 
                 var paramTypes = new List<FormulaType>();
-                foreach (var p in m.GetParameters())
+
+                info._isAsync = m.ReturnType.BaseType == typeof(Task);
+
+                var parmas = m.GetParameters();
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs:215]: >parmas [](http://example.com/codeflow?start=20&length=6)

typo

File: src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs [hunk 6]
```diff
@@ -207,21 +209,36 @@ private FunctionDescr Scan()
...
-                        info._configType = p.ParameterType;
-                    } 
-                    else
+                        info._configType = parmas[i].ParameterType;
+                    }
+                    else if (parmas[i].ParameterType == typeof(CancellationToken) && _info._isAsync)
                     {
-                        // Unknonw parameter type
-                        throw new InvalidOperationException($"Unknown parameter type: {p.Name}, {p.ParameterType}");
+                        throw new InvalidOperationException($"Cancellation token must be the last argument.");
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs:236]: >InvalidOperationException [](http://example.com/codeflow?start=34&length=25)

why do we have this check on both line 236 and line 222?

File: src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs [hunk 7]
```diff
@@ -207,21 +209,36 @@ private FunctionDescr Scan()
...
                 {
-                    if (typeof(FormulaValue).IsAssignableFrom(p.ParameterType))
+                    if (i == parmas.Length - 1 && info._isAsync)
                     {
-                        paramTypes.Add(GetType(p.ParameterType));
-                    } 
-                    else if (p.ParameterType == ConfigType)
+                        if (parmas[i].ParameterType != typeof(CancellationToken))
+                        {
+                            throw new InvalidOperationException($"Last argument must be a cancellation token.");
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs:222]: >"Last argument must be a cancellation token."); [](http://example.com/codeflow?start=65&length=47)

Can we add a test that verifies we hit this?  
Ie, does _not_ have a cancellation token:

Task<StringFormula> Execute();

File: src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs (hunk 6 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs:236]: 222 checks if the last argument is Cancellation token.
while 236 is in-case user passed cancellation token somewhere in middle (in-case they don't know convention) and we want to give the better feedback in error message.

File: src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs (hunk 7 above)
Comment (jas-valgotar) [reply src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs:222]: Added the test

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by MikeStall — APPROVED** [review]
:shipit:

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Pull request merged**